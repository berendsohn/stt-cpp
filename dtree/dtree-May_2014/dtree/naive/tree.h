// -*-C++-*-
// Copyright (c) 2012--2014 David Eisenstat <eisenstatdavid@gmail.com>
//     and Brown University
// Released under http://opensource.org/licenses/MIT
// May 2014 version

#ifndef DTREE_NAIVE_TREE_H_
#define DTREE_NAIVE_TREE_H_

#include "dtree/naive/common.h"

namespace naive {

template<typename Base>
class EndTree : public Base {
 public:
  EndTree() : parent_(NULL) {}
  EndTree* parent() const { return parent_; }
  void set_parent(EndTree* parent) { parent_ = parent; }
  template<typename Forwarder>
  void CopyFrom(const EndTree& node, Forwarder forward);
 private:
  EndTree* parent_;
  DISALLOW_COPY_AND_ASSIGN(EndTree);
};

template<typename Type, typename Base>
class WithStaticValue : public Base {
 public:
  typedef dtree::Nop<Type> Group_;
  Type value() const { return value_; }
  void set_value(Type value) { value_ = value; }
  //
  template<typename Node>
  static Type Value(Node* node);
  template<typename Node>
  static void SetValue(Node* node, Type value);
 protected:
  WithStaticValue() : value_() {}
 private:
  Type value_;
};

template<typename Group, typename Base>
class WithAncValue : public Base {
 public:
  typedef Group Group_;
  typename Group::Type value() const { return value_; }
  void set_value(typename Group::Type value) { value_ = value; }
  void add_to_value(typename Group::Type delta) {
    set_value(Group::Plus(value(), delta));
  }
  void subtract_from_value(typename Group::Type delta) {
    set_value(Group::Minus(value(), delta));
  }
  //
  template<typename Node>
  static typename Group::Type Value(Node* node);
  template<typename Node>
  static void SetValue(Node* node, typename Group::Type value);
  template<typename Node>
  static void AddToAnc(Node* node, typename Group::Type delta);
  template<typename Node>
  static void SubtractFromAnc(Node* node, typename Group::Type delta);
  template<typename Node>
  static void AddToProperAnc(Node* node, typename Group::Type delta);
  template<typename Node>
  static void SubtractFromProperAnc(Node* node, typename Group::Type delta);

 protected:
  WithAncValue() : value_() {}

 private:
  typename Group::Type value_;
};

template<typename Semigroup, typename Base>
class WithAncAggr : public Base {
 public:
  typedef Semigroup Semigroup_;
  typename Semigroup::Type singleton_aggr() const {
    return Semigroup::AggrFromValue(this->value());
  }
  //
  template<typename Node>
  static typename Semigroup::Type AggrAnc(Node* node);
  template<typename Node, typename Predicate>
  static Node* FindDirmostAnc(int dir, Node* node, const Predicate& predicate);
  template<typename Node, typename Predicate>
  static Node* FindLeafmostAnc(Node* node, const Predicate& predicate);
  template<typename Node, typename Predicate>
  static Node* FindRootmostAnc(Node* node, const Predicate& predicate);
  template<typename Node>
  static typename Semigroup::Type AggrProperAnc(Node* node);
  template<typename Node, typename Predicate>
  static Node* FindDirmostProperAnc(int dir,
                                    Node* node,
                                    const Predicate& predicate);
  template<typename Node, typename Predicate>
  static Node* FindLeafmostProperAnc(Node* node, const Predicate& predicate);
  template<typename Node, typename Predicate>
  static Node* FindRootmostProperAnc(Node* node, const Predicate& predicate);

 protected:
  WithAncAggr() {}
};

template<typename Base>
class WithEvertBy : public Base {
 private:
  typedef typename Base::Group_ Group;
 public:
  template<typename Node>
  static void Evert(Node* node);
 protected:
  WithEvertBy() {}
};

template<typename Base>
class WithEvert : public Base {
 public:
  template<typename Node>
  static void Evert(Node* node);
 protected:
  WithEvert() {}
};
}  // namespace naive
#endif  // DTREE_NAIVE_TREE_H_
