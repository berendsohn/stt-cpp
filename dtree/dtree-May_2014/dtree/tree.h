// -*-C++-*-
// Copyright (c) 2012--2014 David Eisenstat <eisenstatdavid@gmail.com>
//     and Brown University
// Released under http://opensource.org/licenses/MIT
// May 2014 version

#ifndef DTREE_TREE_H_
#define DTREE_TREE_H_

#include "dtree/selector.h"

#ifdef NAIVE_DTREE
#include "dtree/naive/tree.h"
#else
#include "dtree/selfadjust/tree.h"
#endif
#endif  // DTREE_TREE_H_
