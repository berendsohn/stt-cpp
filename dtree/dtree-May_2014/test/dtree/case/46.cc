#define DTREE_CONFIG EndTreeWithDesc<WithEvert<WithEvert<WithAncAggr<dtree::DpMin<double>, dtree::DwLessEqual_<double>, WithAncAggr<dtree::DpMin<double>, dtree::DwLessEqual_<double>, WithAncValue<dtree::DpAdd<double>, WithDescAggrOfDescValue<dtree::Count<int>, NullPredicate_, WithAncAggr<dtree::Count<int>, dtree::Index_<int>, Begin > > > > > > > >
#include "test/dtree/tree.cc"
