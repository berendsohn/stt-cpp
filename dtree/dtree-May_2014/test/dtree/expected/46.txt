<<<
  1: vL-11R-3F1 -> 7
  2: vL-2R-7F0
  3: vL-4R-2F1 -> 11
  4: vL14R-8F0
  5: vL9R-2F0 -> 1
  6: vL7R-5F0 -> 7
  7: vL-29R27F1 -> 9
  8: vL-2R-5F0
  9: vL-5R19F1 -> 11
  10: vL-21R24F0
  11: vL3R-2F0 -> 4
  12: vL-6R18F0 -> 6
  13: vL0R0F1 -> 10
  14: vL-2R10F0
  15: vL-13R11F1 -> 1
>>>
<<<
  @2: a1 a1 a1 vL-2R-7F0 aL0R0 aL0R0 f0 f0

      /11: a2 a2 a3 vL-29R32F1 aL-8R0 aL-8R0 f1 f1
       ->3: a1 a1 a1 vL-4R-2F1 aL0R0 aL0R0 f1 f1
        \9: a1 a1 a1 vL-8R21F1 aL0R0 aL0R0 f0 f0
    /7: a5 a5 a9 vL-43R35F1 aL-30R0 aL-30R0 f1 f0
        /12: a1 a1 a1 vL-13R23F0 aL0R0 aL0R0 f1 f1
     ->6: a2 a2 a2 vL7R-5F0 aL-13R0 aL-13R0 f1 f0
        /15: a1 a1 a1 vL14R-2F0 aL0R0 aL0R0 f1 f0
      \1: a2 a2 a3 vL-30R18F0 aL0R-2 aL0R-2 f0 f1
       ->5: a1 a1 a1 vL9R-2F0 aL0R0 aL0R0 f1 f0
  @4: a1 a1 a1 vL14R-8F0 aL0R0 aL0R0 f0 f1

  @8: a1 a1 a1 vL-2R-5F0 aL0R0 aL0R0 f0 f1

    /10: a1 a1 a1 vL24R-21F1 aL0R0 aL0R0 f0 f0
  @13: a2 a2 a2 vL0R0F1 aL0R-21 aL0R-21 f1 f1

  @14: a1 a1 a1 vL-2R10F0 aL0R0 aL0R0 f1 f1
>>>
